### PR TITLE
Navigating a cross-origin iframe to the same URL should not replace the current HistoryItem

### PR DESCRIPTION
**Pull request by cdumez**
#### 778e351efb2f883b957255017712151e3fffa629
<pre>
Navigating a cross-origin iframe to the same URL should not replace the current HistoryItem
<a href="https://bugs.webkit.org/show_bug.cgi?id=245246">https://bugs.webkit.org/show_bug.cgi?id=245246</a>

Reviewed by Darin Adler.

Navigating a cross-origin iframe to the same URL should not replace the current HistoryItem.
This aligns our behavior with Blink and Gecko.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/navigate-cross-origin-iframe-to-same-url-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/navigate-cross-origin-iframe-to-same-url-with-fragment-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadArchive):
(WebCore::FrameLoader::loadInSameDocument):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::load):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::continueFragmentScrollAfterNavigationPolicy):
(WebCore::FrameLoader::shouldTreatURLAsSameAsCurrent const):
(WebCore::FrameLoader::loadSameDocumentItem):
* Source/WebCore/loader/FrameLoader.h:

Canonical link: <a href="https://commits.webkit.org/254563@main">https://commits.webkit.org/254563@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/1effffcd67b6e2a515316be651a631b561d2e6dc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89466 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34020 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/98786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/155092 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93474 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32525 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81826 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93198 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95113 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/25838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/25767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/80716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/68756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30284 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30022 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3202 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33469 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32179 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | |

### CODE REVIEW

File: Source/WebCore/loader/FrameLoader.cpp [hunk 1]
```diff
@@ -1622,8 +1622,11 @@ void FrameLoader::loadWithDocumentLoader(DocumentLoader* loader, FrameLoadType t
         oldDocumentLoader->setLastCheckedRequest(ResourceRequest());
         policyChecker().stopCheck();
         RELEASE_ASSERT(!isBackForwardLoadType(policyChecker().loadType()) || history().provisionalItem());
-        policyChecker().checkNavigationPolicy(ResourceRequest(loader->request()), ResourceResponse { }  /* redirectResponse */, oldDocumentLoader.get(), WTFMove(formState), [this, protectedFrame = Ref { m_frame }] (const ResourceRequest& request, WeakPtr<FormState>&&, NavigationPolicyDecision navigationPolicyDecision) {
-            continueFragmentScrollAfterNavigationPolicy(request, navigationPolicyDecision == NavigationPolicyDecision::ContinueLoad);
+        RefPtr<SecurityOrigin> requesterOrigin;
+        if (auto& requester = loader->triggeringAction().requester())
+            requesterOrigin = requester->securityOrigin.copyRef();
```
Comment (darinadler) [inline Source/WebCore/loader/FrameLoader.cpp]: Kind of surprised we have to write `copyRef` here. I thought we could do `RefPtr` = `Ref` without any function call.

**Review by darinadler — APPROVED** [review]

File: Source/WebCore/loader/FrameLoader.cpp (hunk 1 above)
Comment (cdumez) [reply Source/WebCore/loader/FrameLoader.cpp]: I will try. I think I assumed this wouldn't work without trying.

Comment (cdumez) [reply Source/WebCore/loader/FrameLoader.cpp]: It doesn't currently work:
loader/FrameLoader.cpp:1627:29: error: no viable overloaded '='
            requesterOrigin = requester->securityOrigin;

Comment (darinadler) [reply Source/WebCore/loader/FrameLoader.cpp]: I see now that we have the `RefPtr(Ref&&)` constructor but not the `RefPtr(const Ref&)` constructor. Not sure that’s really what we want in the future.

**Review by darinadler — APPROVED** [review]

File: Source/WebCore/loader/FrameLoader.cpp (hunk 1 above)
Comment (cdumez) [reply Source/WebCore/loader/FrameLoader.cpp]: It is indeed a bit annoying to have to copyRef() here. However, one benefit I guess is that it encourages developers to WTFMove() whenever possible.

Comment (darinadler) [reply Source/WebCore/loader/FrameLoader.cpp]: Yes, that was the original reasoning. But we don’t do that for copying a `RefPtr` to another `RefPtr` and we no longer do it for copying a `Ref` to a `Ref`, which we used to do, so it seems peculiar to still do that for `Ref` to `RefPtr`.

**Pull request closed without merging**